src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs

```diff
@@ -0,0 +1,288 @@
+        /// <param name="remoteEndPoint">The connection's <see cref="ConnectionBase.RemoteEndPoint"/>.</param>
+        /// <returns>A new <see cref="Connection"/>.</returns>
+        public static Connection FromPipe(IDuplexPipe pipe, IConnectionProperties? properties = null, EndPoint? localEndPoint = null, EndPoint? remoteEndPoint = null)
+        {
+            if (pipe == null) throw new ArgumentNullException(nameof(pipe));
+            return new ConnectionFromPipe(pipe, disposable: null, properties, localEndPoint, remoteEndPoint);
+        }
+
+        /// <summary>
+        /// Creates a connection for an <see cref="IDuplexPipe"/>.
+        /// </summary>
+        /// <param name="pipe">The connection's <see cref="Connection.Pipe"/>.</param>
+        /// <param name="disposable">An <see cref="IDisposable"/> to call when the connection is being closed.</param>
+        /// <param name="properties">The connection's <see cref="ConnectionBase.ConnectionProperties"/>.</param>
+        /// <param name="localEndPoint">The connection's <see cref="ConnectionBase.LocalEndPoint"/>.</param>
+        /// <param name="remoteEndPoint">The connection's <see cref="ConnectionBase.RemoteEndPoint"/>.</param>
+        /// <returns>A new <see cref="Connection"/>.</returns>
+        public static Connection FromPipe(IDuplexPipe pipe, IDisposable disposable, IConnectionProperties? properties = null, EndPoint? localEndPoint = null, EndPoint? remoteEndPoint = null)
+        {
+            if (pipe == null) throw new ArgumentNullException(nameof(pipe));
+            return new ConnectionFromPipe(pipe, disposable, properties, localEndPoint, remoteEndPoint);
+        }
+
+        /// <summary>
+        /// Creates a connection for an <see cref="IDuplexPipe"/>.
+        /// </summary>
+        /// <param name="pipe">The connection's <see cref="Connection.Pipe"/>.</param>
+        /// <param name="disposable">An <see cref="IAsyncDisposable"/> to call when the connection is being closed.</param>
+        /// <param name="properties">The connection's <see cref="ConnectionBase.ConnectionProperties"/>.</param>
+        /// <param name="localEndPoint">The connection's <see cref="ConnectionBase.LocalEndPoint"/>.</param>
+        /// <param name="remoteEndPoint">The connection's <see cref="ConnectionBase.RemoteEndPoint"/>.</param>
+        /// <returns>A new <see cref="Connection"/>.</returns>
+        public static Connection FromPipe(IDuplexPipe pipe, IAsyncDisposable disposable, IConnectionProperties? properties = null, EndPoint? localEndPoint = null, EndPoint? remoteEndPoint = null)
+        {
+            if (pipe == null) throw new ArgumentNullException(nameof(pipe));
+            return new ConnectionFromPipe(pipe, disposable, properties, localEndPoint, remoteEndPoint);
+        }
+
+        private sealed class ConnectionFromStream : Connection, IConnectionProperties
+        {
+            private new Stream? _stream;
```

Comment:
Can't use the Connection._stream as you might have this happen --

```c#
_pipe = CreatePipe() -> default implementation wraps CreateStream().
```

So I've renamed to `_originalStream` and `_originalPipe`.

Honestly it's weird that it is considered a shadowing when the shadowed variable is private and not so visible to that code :/

Reply:
It is shadowing because these are nested types. Nested types have access to private fields/methods of the parent.

If you moved the private types out of the parent type then it would no longer be shadowing.